petstore/src/main/scala/io/finch/petstore/endpoint.scala

```diff
@@ -0,0 +1,238 @@
+import io.finch.route._
+
+/**
+ * Provides the paths and endpoints for all the API's public service methods.
+ */
+object endpoint extends ErrorHandling {
+
+  /**
+   * Private method that compiles all pet service endpoints.
+   * @return Bundled compilation of all pet service endpoints.
+   */
+  private def petEndpts(db: PetstoreDb) =
+    getPetEndpt(db) :+:
+    addPetEndpt(db) :+:
+    updatePetEndpt(db) :+:
+    getPetsByStatusEndpt(db) :+:
+    findPetsByTagEndpt(db) :+:
+    deletePetEndpt(db) :+:
+    updatePetViaFormEndpt(db) :+:
+    uploadImageEndpt(db)
+
+  /**
+   * Private method that compiles all store service endpoints.
+   * @return Bundled compilation of all store service endpoints.
+   */
+  private def storeEndpts(db: PetstoreDb) =
+    getInventoryEndpt(db) :+:
+    addOrderEndpt(db) :+:
+    deleteOrderEndpt(db) :+:
+    findOrderEndpt(db)
+
+  /**
+   * Private method that compiles all user service endpoints.
+   * @return Bundled compilation of all user service endpoints.
+   */
+  private def userEndpts(db: PetstoreDb) =
+    addUserEndpt(db) :+:
+    addUsersViaList(db) :+:
+    addUsersViaArray(db) :+:
+    getUserEndpt(db) :+:
+    deleteUserEndpt(db) :+:
+    updateUserEndpt(db)
+
+  /**
+   * Compiles together all the endpoints relating to public service methods.
+   * @return A service that contains all provided endpoints of the API.
+   */
+  def makeService(db: PetstoreDb): Service[Request, Response] = handleExceptions andThen (
+    petEndpts(db) :+:
+    storeEndpts(db) :+:
+    userEndpts(db)
+  ).toService
+
+  /**
+   * The long passed in the path becomes the ID of the Pet fetched.
+   * @return A Router that contains the Pet fetched.
+   */
+  def getPetEndpt(db: PetstoreDb): Router[Future[Pet]] = Get / "pet" / long /> {petId: Long =>
+    db.getPet(petId)
+  }
+
+  /**
+   * The pet to be added must be passed in the body.
+   * @return A Router that contains a RequestReader of the ID of the Pet added.
+   */
+  def addPetEndpt(db: PetstoreDb): Router[RequestReader[Long]] = Post / "pet" />{
+    body.as[Pet].embedFlatMap {pet =>
+      for{
+        realId <- db.addPet(pet)
+      }yield realId
+    }
+  }
+
+  /**
+   * The updated, better version of the current pet must be passed in the body.
+   * @return A Router that contains a RequestReader of the updated Pet.
+   */
+  def updatePetEndpt(db: PetstoreDb): Router[RequestReader[Pet]] = Put / "pet" /> {
+    body.as[Pet].embedFlatMap { pet =>
+      for {
+        _ <- db.updatePet(pet.copy(id = Some(pet.id.getOrElse(-1))))
+        newPet <- db.getPet(pet.id.getOrElse(-1))
+      } yield newPet
+    }
+  }
+
+  /**
+   * The status is passed as a query parameter.
+   * @return A Router that contains a RequestReader of the sequence of all Pets with the Status in question.
+   */
+  def getPetsByStatusEndpt(db: PetstoreDb): Router[RequestReader[Seq[Pet]]] = Get / "pet" / "findByStatus" />{
+    (reader.statusReader).embedFlatMap{
```

Comment:
It should be possible to write `reader.statusReader.embedFlatMap(db.getPetsByStatus)`.